### PR TITLE
Adding calibration and registration tools to examples.

### PR DESCRIPTION
**Pull request by kyrendon-msft**
### Description of the changes:
- Adding python code leveraging OpenCV to calibrate any given camera.
- Adding python code to perform registration (extrinsics) between two cameras, also reports registration error.
- Adding example data for users.
- Adding a README.md to describe how to use the tools
- Adding three different Charuco templates for users, with example data using the mid size one.
- Output a json with calibration and registration data between cameras.

### Before submitting a Pull Request:
- [x] I reviewed [CONTRIBUTING.md](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/CONTRIBUTING.md)
- [x] I [built my changes](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/building.md) locally
- [x] I ran the [unit tests](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/testing.md)
- [x] I ran the [functional tests](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/testing.md) with a device
- [x] I ran the [performance tests](https://github.com/Microsoft/Azure-Kinect-Sensor-SDK/blob/develop/docs/testing.md) with a device

### I tested changes on: 
- [x] Windows
- [x] Linux

### CODE REVIEW

**Comment by seigeweapon** [comment]
hi, great to see this, a much needed function!

I wonder what's the accuracy like? Is the calibrated parameters close to the factory calibration?

**Review by resuther-msft — DISMISSED** [review]
Everything looks good.

**Comment by resuther-msft** [comment]
> hi, great to see this, a much needed function!
> 
> I wonder what's the accuracy like? Is the calibrated parameters close to the factory calibration?

@seigeweapon The accuracy of the calibration from this tool is as good, if not better than the factory calibration. You can use this tool to calibrate a device, and then use the tools in the examples/depth_eval_tools folder to evaluate the calibration quality for the depth sensor.

File: examples/calibration_registration/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        ![Good Image](example_files/registration/color-0.jpg "Good Image")
+   6. For registration, this means taking images from both cameras with a board that is not moved between captures. Unlike calibration, this requires only one capture per camera.
+
+2. How to calibrate a camera.
+   1. Calibration typically requires about 30 images and relies on the Brown Conrady distortion model.
+   2. The script `calibrate.py` is designed to take in command line arguments so it can be used personally collected data without modification. For further details please see [Camera Calibration](https://docs.opencv.org/master/dc/dbb/tutorial_py_calibration.html) in the OpenCV documentation.
+   3. If you already have a collected set of data, run the calibration code this way:
+
+      ``` windows
+      python <full_path>\calibrate.py. –dataset-dir <dataset_dir> --template <full_path_tempate>
```
Comment (JonathanESantos) [inline examples/calibration_registration/README.md]: There is an extra period after calibrate.py.
I believe it is --dataset-dir (double dash).
full_path_tempate is misspelled (template missing the "L").

File: examples/calibration_registration/README.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      ```
+
+      The tool may optinally take in an existing calibration file with the flag `--calib-file`
```
Comment (JonathanESantos) [inline examples/calibration_registration/README.md]: spelling: optionally

File: examples/calibration_registration/README.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+   1. This requires that both cameras have been successfully calibrated prior to this tool.
+   2. Find the five files you will need. Two images (one per camera) taken at the same time and the corresponding `calib.yml` calibration files generated for a well calibrated camera. Also pass in the template file.
+        * Example: RGB and IR images taken simultaneously.
+        * ![Color Image](example_files/registration/color-0.jpg "Color Image")
+        * ![IR Image](example_files\registration/ir-0.png "IR Image")
+   3. If you already have a collected set of data, run the calibration code this way:
+
+        ``` windows
+        python <full_path>\register.py --img-a <full_path_img_a> --img-b <full_path_img_b> `
+        --template <full_path_tempate> --calib-a <full_path_calibration_a> `
```
Comment (JonathanESantos) [inline examples/calibration_registration/README.md]: full_path_tempate has "template" misspelled.

File: examples/calibration_registration/README.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        --calib-b <full_path_calibration_b> --out-dir <full_path_calibration_b>
```
Comment (JonathanESantos) [inline examples/calibration_registration/README.md]: --out-dir <full_path_calibration_b> is a little confusing, seems like "--out-dir <output_directory>" might be a more appropriate choice of words...unless the output needs something from the full_path_calibration_b directory, then it makes sense.

File: examples/calibration_registration/camera_tools.py [hunk 5]
```diff
@@ -0,0 +1,748 @@
...
+      rms: the overall RMS re-projection error in pixels.
+      k_matrix: camera matrix.
+      dist: Lens distortion coeffs. OpenCV model with 8 distortion is equivalent
+        to Brown-Conrady model. (used in K4A).
+      img_size: [width; height].
+      rvecs: list of rotation vectors.
+      tvecs: list of translation vectors.
+  """
+
+  # Check valididty of initial calibration file.
```
Comment (JonathanESantos) [inline examples/calibration_registration/camera_tools.py]: sp: valididty

File: examples/calibration_registration/camera_tools.py [hunk 6]
```diff
@@ -0,0 +1,748 @@
...
+                           k_matrix,
+                           dist,
+                           flags=flags,
+                           criteria=criteria)
+
+  # Check Quality of each image.
+  n_low_rms = [err[0] for err in perViewErrors if err[0] <= per_view_threshold]
+  num_good_images = len(n_low_rms)
+
+  # Report which indexes are failing in perViewErros.
```
Comment (JonathanESantos) [inline examples/calibration_registration/camera_tools.py]: sp: perViewErros (missing r in Errors)

File: examples/calibration_registration/camera_tools.py [hunk 7]
```diff
@@ -0,0 +1,748 @@
...
+      tmat_b_to_a: Numpy array of the translation vector from B to A.
+      rms: Reprojection error expressed as Root mean square of pixel diffs
+        between markers.
+  """
+
+  # File exists checks.
+  if not os.path.exists(img_a):
+    raise FileExistsError(f"Image file not found for camera A @ {img_a}")
+  if not os.path.exists(img_b):
+    raise FileExistsError(f"Iamge file not found for camera B @ {img_b}")
```
Comment (JonathanESantos) [inline examples/calibration_registration/camera_tools.py]: sp: Iamge

**Review by JonathanESantos — DISMISSED** [review]
Approved with minor spelling mistakes.

**Comment by kyrendon-msft** [comment]
> hi, great to see this, a much needed function!
> 
> I wonder what's the accuracy like? Is the calibrated parameters close to the factory calibration?

Yes, this calibration is as good as factory.

File: examples/calibration_registration/camera_tools.py (hunk 7 above)
Comment (kyrendon-msft) [reply examples/calibration_registration/camera_tools.py]: fixed.

File: examples/calibration_registration/README.md (hunk 4 above)
Comment (kyrendon-msft) [reply examples/calibration_registration/README.md]: Thanks for the catch. There should be no expectation of existing files, so it should be `out_dir`

**Pull request merged**